lib/backend.js

```diff
@@ -95,7 +94,10 @@ Backend.prototype.SNAPSHOT_TYPES = {
 Backend.prototype._shimDocAction = function() {
   if (warnDeprecatedDoc) {
     warnDeprecatedDoc = false;
-    console.warn('DEPRECATED: "doc" middleware action. Use "readSnapshots" instead. Pass `disableDocAction: true` option to ShareDB to disable the "doc" action and this warning.');
+    console.warn([
+      'DEPRECATED: "doc" middleware action. Use "readSnapshots" instead.',
+      'Pass `disableDocAction: true` option to ShareDB to disable the "doc" action and this warning.'
+    ].join(' '));
```

Comment:
To reduce memory allocation overhead from string joining, I'd prefer this to either:
* Remain as one long inline string
* Or do the join once at module level, instead of inside the function

Same thing below for `_shimAfterSubmit`.

Reply:
Let's move this to the top of the file

```js
"with this " +
"style"
```